services/stats/stats.go

```diff
@@ -93,327 +78,190 @@ func (t Tags) String() string {
 	})
 }
 
-// NewStat creates a new RudderStats with provided Name and Type
-func (s *HandleT) NewStat(Name, StatType string) (rStats RudderStats) {
-	return &RudderStatsT{
-		Name:     Name,
-		StatType: StatType,
-		Client:   client,
-	}
-}
-
-func (s *HandleT) NewTaggedStat(Name, StatType string, tags Tags) (rStats RudderStats) {
-	return newTaggedStat(Name, StatType, tags, 1)
-}
-
-func (s *HandleT) NewSampledTaggedStat(Name, StatType string, tags Tags) (rStats RudderStats) {
-	return newTaggedStat(Name, StatType, tags, statsSamplingRate)
-}
-
-func CleanupTagsBasedOnDeploymentType(tags Tags, key string) Tags {
-	deploymentType, err := deployment.GetFromEnv()
-	if err != nil {
-		pkgLogger.Errorf("error while getting deployment type: %w", err)
-		return tags
-	}
-
-	if deploymentType == deployment.MultiTenantType {
-		isNamespaced := config.IsSet("WORKSPACE_NAMESPACE")
-		if !isNamespaced || (isNamespaced && strings.Contains(config.GetString("WORKSPACE_NAMESPACE", ""), "free")) {
-			delete(tags, key)
+func (s *statsdStats) getNewStatsdClientWithExpoBackoff(ctx context.Context, opts ...statsd.Option) (*statsd.Client, error) {
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = time.Minute
+	bo.MaxElapsedTime = 0
+	boCtx := backoff.WithContext(bo, ctx)
+	var err error
+	var c *statsd.Client
+	op := func() error {
+		c, err = statsd.New(opts...)
+		if err != nil {
+			s.log.Errorf("error while setting statsd client: %v", err)
 		}
+		return err
 	}
 
-	return tags
+	err = backoff.Retry(op, boCtx)
+	return c, err
 }
 
-func newTaggedStat(Name, StatType string, tags Tags, samplingRate float32) (rStats RudderStats) {
-	// If stats is not enabled, returning a dummy struct
-	if !statsEnabled {
-		return &RudderStatsT{
-			Name:        Name,
-			StatType:    StatType,
-			Client:      nil,
-			dontProcess: true,
-		}
-	}
-
-	// Clean up tags based on deployment type. No need to send workspace id tag for free tier customers.
-	tags = CleanupTagsBasedOnDeploymentType(tags, "workspaceId")
-
-	if tags == nil {
-		tags = make(Tags)
-	}
-	// key comprises of the measurement name plus all tag-value pairs
-	taggedClientKey := StatType + tags.String()
-
-	taggedClientsMapLock.RLock()
-	taggedClient, found := taggedClientsMap[taggedClientKey]
-	taggedClientsMapLock.RUnlock()
-
-	if !found {
-		taggedClientsMapLock.Lock()
-		if taggedClient, found = taggedClientsMap[taggedClientKey]; !found { // double check for race
-			tagVals := tags.Strings()
-			if !connEstablished {
-				taggedClientPendingTags = append(taggedClientPendingTags, taggedClientKey)
-				taggedClientPendingKeys = append(taggedClientPendingKeys, tagVals)
-			}
-			taggedClient = client.Clone(conn, statsd.TagsFormat(getTagsFormat()), defaultTags(), statsd.Tags(tagVals...), statsd.SampleRate(samplingRate))
-			taggedClientsMap[taggedClientKey] = taggedClient
+func (s *statsdStats) collectPeriodicStats() {
+	gaugeFunc := func(key string, val uint64) {
+		if s.state.client.statsd != nil {
```

Comment:
Do we need this nil check?

Reply:
yes, `statsd` client might be nil in case of errors. But I will replace it with `s.state.client.ready()`

Reply:
`statsd` client will never be nil right?

Reply:
statsd client will be nil if `Stats` is not [started](https://github.com/rudderlabs/rudder-server/blob/a8b1ab147a3264b9038000ff107f20dd554f8bf7/services/stats/stats.go#L117). This is how all tests, other than stats’ own tests, run: stats is initialized but not started

Reply:
but you are correct, this specific function will only be called after a statsd client has been successfully created...